### PR TITLE
feat(element_injector): add distance to propertly implement @parent

### PR DESCRIPTION
**Pull request by vsavkin**
@tbosch can you review this PR? It is needed to implement emulated shadow DOM.

### CODE REVIEW

File: modules/core/src/compiler/pipeline/compile_element.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ export class CompileElement {
   inheritedProtoView:ProtoView;
   inheritedProtoElementInjector:ProtoElementInjector;
   inheritedElementBinder:ElementBinder;
+  distanceToParentInjector:int;
```
Comment (gkalpak) [inline modules/core/src/compiler/pipeline/compile_element.js]: Just out of curiosity: Is there any reason why this is of type `int`, whereas `distanceToParent` above is of type `number` ?

File: modules/core/src/compiler/pipeline/proto_element_injector_builder.js [hunk 2]
```diff
@@ -39,13 +41,21 @@ export class ProtoElementInjectorBuilder extends CompileStep {
...
       var protoView = current.inheritedProtoView;
       var hasComponent = isPresent(current.componentDirective);
-      inheritedProtoElementInjector = this.internalCreateProtoElementInjector(
-        parentProtoElementInjector, protoView.elementBinders.length, injectorBindings, hasComponent
+
+      current.inheritedProtoElementInjector = this.internalCreateProtoElementInjector(
+        parentProtoElementInjector, protoView.elementBinders.length, injectorBindings,
+        hasComponent, distanceToParentInjector
       );
+      current.distanceToParentInjector = 0;
```
Comment (vicb) [inline modules/core/src/compiler/pipeline/proto_element_injector_builder.js]: Is is intended to set `distanceToParentInjector` after the injector has been created or could `this.internalCreateProtoElementInjector` be called with a distance set set 0 ?

File: modules/core/src/compiler/element_injector.js [hunk 3]
```diff
@@ -411,8 +416,8 @@ export class ElementInjector extends TreeNode {
       var dir = ei._getDirectiveByKeyId(key.id);
       if (dir !== _undefined) return dir;
 
+      depth -= ei._proto.distanceToParent;
```
Comment (vicb) [inline modules/core/src/compiler/element_injector.js]: should the code l 402 also be modified to take into account the distance to parent ?

```
     if (! this._shouldIncludeSelf(depth)) {
       depth -= 1;
       ei = ei._parent;
     }
```

File: modules/core/test/compiler/element_injector_spec.js [hunk 4]
```diff
@@ -144,6 +144,30 @@ export function main() {
           [c2, 'child2']
         ])).toEqual(["parent", ["child1", "child2"]]);
       });
+
+      describe("direct parent", () => {
+        it("returns parent injector when distance is 1", () => {
```
Comment (vicb) [inline modules/core/test/compiler/element_injector_spec.js]: "should return" here and below

File: modules/core/src/compiler/element_injector.js [hunk 5]
```diff
@@ -165,9 +165,10 @@ export class ProtoElementInjector  {
   parent:ProtoElementInjector;
   index:int;
   view:View;
-  constructor(parent:ProtoElementInjector, index:int, bindings:List, firstBindingIsComponent:boolean = false) {
+  constructor(parent:ProtoElementInjector, index:int, bindings:List, firstBindingIsComponent:boolean = false, distanceToParent:number = 0) {
     this.parent = parent;
     this.index = index;
+    this.distanceToParent = distanceToParent;
```
Comment (vicb) [inline modules/core/src/compiler/element_injector.js]: `distanceToParent` should be declared (causes Travis to fail for Dart)

**Comment by vicb** [comment]
LGTM with a few comments / questions

File: modules/core/src/compiler/pipeline/compile_element.js (hunk 1 above)
Comment (vsavkin) [inline modules/core/src/compiler/pipeline/compile_element.js]: It should be `number`. At some point we were thinking to support `int` as an alias for `number`, but not any more.

**Pull request merged**